Lib/collections/__init__.py

```diff
@@ -390,46 +340,109 @@ def namedtuple(typename, field_names, *, verbose=False, rename=False, module=Non
-    class_definition = _class_template.format(
-        typename = typename,
-        field_names = tuple(field_names),
-        num_fields = len(field_names),
-        arg_list = repr(tuple(field_names)).replace("'", "")[1:-1],
-        repr_fmt = ', '.join(_repr_template.format(name=name)
-                             for name in field_names),
-        field_defs = '\n'.join(_field_template.format(index=index, name=name)
-                               for index, name in enumerate(field_names))
+    # Variables used in the methods and docstrings
+    field_names = tuple(field_names)
+    num_fields = len(field_names)
+    arg_list = repr(field_names).replace("'", "")[1:-1]
+    repr_fmt = '(' + ', '.join(f'{name}=%r' for name in field_names) + ')'
+    module_name = f'namedtuple_{typename}'
+    tuple_new = tuple.__new__
+    _len = len
+
+    # Create all the named tuple methods to be added to the class namespace
+
+    s = f'def __new__(_cls, {arg_list}): return _tuple_new(_cls, ({arg_list}))'
+    namespace = dict(_tuple_new=tuple_new, __name__=module_name)
+    # Note: exec() has the side-effect of interning the typename and field names
+    exec(s, namespace)
+    __new__ = namespace['__new__']
+    __new__.__doc__ = f'Create new instance of {typename}({arg_list})'
+
+    @classmethod
+    def _make(cls, iterable):
+        result = tuple_new(cls, iterable)
+        if _len(result) != num_fields:
+            raise TypeError(f'Expected {num_fields} arguments, got {len(result)}')
+        return result
+
+    _make.__func__.__doc__ = (f'Make a new {typename} object from a sequence '
+                              'or iterable')
+
+    def _replace(_self, **kwds):
+        result = _self._make(map(kwds.pop, field_names, _self))
+        if kwds:
+            raise ValueError(f'Got unexpected field names: {list(kwds)!r}')
+        return result
+
+    _replace.__doc__ = (f'Return a new {typename} object replacing specified '
+                        'fields with new values')
+
+    def __repr__(self):
+        'Return a nicely formatted representation string'
+        return self.__class__.__name__ + repr_fmt % self
+
+    def _asdict(self):
+        'Return a new OrderedDict which maps field names to their values.'
+        return OrderedDict(zip(self._fields, self))
+
+    def __getnewargs__(self):
+        'Return self as a plain tuple.  Used by copy and pickle.'
+        return tuple(self)
+
+    # Modify function metadata to help with introspection and debugging
+
+    for method in (__new__, _make.__func__, _replace,
+                   __repr__, _asdict, __getnewargs__):
+        method.__module__ = module_name
+        method.__qualname__ = f'{typename}.{method.__name__}'
+
+    # Helper function used in the class creation
+
+    def reuse_itemgetter(index):
+        try:
+            return _nt_itemgetters[index]
+        except KeyError:
+            getter = _nt_itemgetters[index] = _itemgetter(index)
+            return getter
```

Comment:
How about caching property?
```py
doc = f'Alias for field number {index}'
prop = _nt_props[index] = property(_itemgetter(index), doc=doc)
return prop
```

Reply:
Can't cache property() the docstring is mutable and can updated by the user, so the property objects need to be distinct.

Reply:
Oh I didn't realized it.
Anyway, I prefer caching docstring to `sys.intern`, because `sys.intern` approach
requires temporal new string and hashing it.
But that's just my preference and there are no real problem.